### PR TITLE
nvprims bookend non compute

### PR DESCRIPTION
**Pull request by jjsjann123**
Cherry-pickeding: https://github.com/csarofeen/pytorch/pull/2099

1. enabling bookend non-compute-ops pass on nvfuser
2. fixing bookend op check on intermediate tensor as partition inputs
3. python tests added for: `getitem` special handling bookend_non_compute removal
4. patching dfs by excluding dfs within partition to avoid going over recursion limitation

### CODE REVIEW

**Comment by jjsjann123** [comment]
Note that I also somewhat patched #87008 here. slightly different implementation. I'll switch it to iterative traversal instead in a later PR.

**Comment by mruberry** [comment]
@SherlockNoMad would you review this?

**Comment by SherlockNoMad** [comment]
Just curious, what's the problem for nvfuser having non-compute node at the boundary of partition?

File: torch/fx/passes/infra/partitioner.py [hunk 1]
```diff
@@ -205,6 +208,62 @@ def fuse_partitions(self, partitions: List[Partition]) -> GraphModule:
...
+        for partition in partitions:
+            # Note it's ok to use `set` here, since we are only query if a node
+            # has been removed. We are NEVER going to iterate on nodes inside
+            # the set.
+            remove_node: Set[Node] = set()
+            for node in partition.nodes:
+                if is_non_compute_node(node) and \
+                    (is_transparent_input_node(node, partition.nodes, remove_node) or
+                     is_transparent_output_node(node, partition.nodes, remove_node)):
+                    remove_node.add(node)
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/partitioner.py:262]: After removing nodes, you could end up as a single node partition? if allows_single_node_partition=False, you need to clean up such one-node partition.

**Review by SherlockNoMad — APPROVED** [review]
Per comment.

**Comment by jjsjann123** [comment]
> Just curious, what's the problem for nvfuser having non-compute node at the boundary of partition?

Not a *problem* per se. It's mostly a temporary WAR to avoid bookend non-compute node gives us tricky kernel generation. At this moment this helps us avoid bad perf in some cases in torchbench. :stuck_out_tongue:

File: torch/fx/passes/infra/partitioner.py (hunk 1 above)
Comment (jjsjann123) [reply torch/fx/passes/infra/partitioner.py:262]: https://github.com/pytorch/pytorch/blob/02c1a304fa801942258a15a7e50abaa92aca2ddf/torch/fx/passes/infra/partitioner.py#L181-L193

^^^ non-compute node is already being accommodated in single node partition. So this isn't a problem.

**Comment by IvanYashchuk** [comment]
> > Just curious, what's the problem for nvfuser having non-compute node at the boundary of partition?
> 
> Not a _problem_ per se. It's mostly a temporary WAR to avoid bookend non-compute node gives us tricky kernel generation. At this moment this helps us avoid bad perf in some cases in torchbench. 😛

If a non-compute node is the output of a fusion group then unnecessary copies are made. Here's an example:
```py
# when translated to nvFuser
# this function would create a kernel with two output tensors not connected to each other
# resulting in two memory operations
def func1(a, b):
    c = a.cos() + b.sin()
    return c, c.unsqueeze(0)

# but here's it's fine because unsqueeze is happening outside of nvFuser
# and in PyTorch eager it's a metadata manipulation with no memory movements
def func2(a, b):
    c = nvfuser_add_cos_sin(a, b)
    return c, c.unsqueeze(0)

**Pull request closed without merging**